lib/boundkeypair/join_state.go

```diff
@@ -0,0 +1,216 @@
+	"time"
+
+	"github.com/go-jose/go-jose/v3"
+	"github.com/go-jose/go-jose/v3/jwt"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils/keys"
+	libjwt "github.com/gravitational/teleport/lib/jwt"
+)
+
+// JoinState is a signed JWT stored on joining clients alongside their usual
+// certificate bundle, used as an additional layer of verification for
+// subsequent join attempts.
+type JoinState struct {
+	*jwt.Claims
+
+	// BotInstanceID is the bot instance ID associated with this join state when
+	// it was generated. A new bot instance may be created during the joining
+	// process if the previous instance expired. The
+	// `(bot_instance_id, recovery_sequence)` tuple is considered functionally
+	// unique to each issued join state for verification purposes, and will
+	// remain unchanged until the next successful recovery.
+	BotInstanceID string `json:"bot_instance_id"`
+
+	// RecoverySequence is the recovery sequence number. This is incremented
+	// each time a recovery is performed, including on first join. This counter
+	// is not reset if a new bot instance is generated.
+	RecoverySequence uint32 `json:"recovery_sequence"`
+
+	// RecoveryLimit is the maximum number of recovery attempts allowed as of
+	// the time this join state was issued. This field is informational, and is
+	// expected to be modified server-side to allow additional joins once the
+	// limit is reached.
+	RecoveryLimit uint32 `json:"recovery_limit"`
+
+	// RecoveryMode is the currently configured recovery mode set in
+	// `spec.bound_keypair.recovery.mode`. This field is informational; clients
+	// may opt to use this and the recovery limit to, e.g., generate a warning
+	// if recovery limits are enforced and the remaining attempts are below some
+	// threshold.
+	RecoveryMode string `json:"recovery_mode"`
+}
+
+// JoinStateParams contains parameters for issuing and verifying join state
+// JWTs.
+type JoinStateParams struct {
+	Clock clockwork.Clock
+
+	ClusterName string
+	Token       *types.ProvisionTokenV2
+}
+
+// IssueJoinState generates a join state document from the provided token and
+// returns a compact serialized, signed JWT. The token must be up-to-date at the
+// time of issuance, i.e. the recovery count must have been incremented already.
+func IssueJoinState(signer crypto.Signer, params *JoinStateParams) (string, error) {
+	spec := params.Token.Spec.BoundKeypair
+	if spec == nil {
+		return "", trace.BadParameter("spec.bound_keypair: required field is missing")
+	}
+
+	if params.Token.Status == nil || params.Token.Status.BoundKeypair == nil {
+		return "", trace.BadParameter("status.bound_keypair: required field is missing")
+	}
+
+	status := params.Token.Status.BoundKeypair
+
+	if params.Token.Spec.BotName == "" {
+		return "", trace.BadParameter("spec.bot_name: required field is empty")
+	}
+
+	state := &JoinState{
+		Claims: &jwt.Claims{
```

Comment:
I think we should call out explicitly why there is no expiry here - since I think that'll be unusual enough to raise eyebrows.